test/matrix.lean

```diff
@@ -50,4 +52,25 @@ example {a b c d e f g h : α} : ![a, b, c, d, e, f, g, h] 7 = h := by simp
 example {a b c d e f g h : α} : ![a, b, c, d, e, f, g, h] 37 = f := by simp
 example {a b c d e f g h : α} : ![a, b, c, d, e, f, g, h] 99 = d := by simp
 
+example {α : Type} [comm_ring α] {a b c d : α} :
+  matrix.det ![![a, b], ![c, d]] = a * d - b * c :=
+begin
+  -- TODO: can we make this require less steering?
+  simp [matrix.det, finset.univ_perm_fin_succ, ←finset.univ_product_univ, finset.sum_product,
+        fin.sum_univ_succ, fin.prod_univ_succ],
+  ring
+end
+
+example {R : Type*} [comm_ring R] (A : matrix (fin 3) (fin 3) R) {a b c d e f g h i}
+        (hA : A = ![![a, b, c], ![d, e, f], ![g, h, i]]) :
+        matrix.det A = a * e * i - a * f * h - b * d * i + b * f * g + c * d * h - c * e * g :=
+begin
+  -- TODO: write swap simp lemmas
+  have h0 : equiv.swap (0 : fin 3) 1 (fin.succ 1) = 2 := dec_trivial,
+  have h1 : equiv.swap (0 : fin 3) (fin.succ 1) 1 = 1 := dec_trivial,
+  simp [h0, h1, hA, matrix.det, finset.univ_perm_fin_succ, ←finset.univ_product_univ,
+        finset.sum_product, fin.sum_univ_succ, fin.prod_univ_succ, fin.succ_ne_zero],
+  ring
+end
```

Comment:
These proofs are not things that someone like me is going to be able to find themselves, and nobody will think to look in `test`. Why not just put them in mathlib?

Reply:
The idea behind these proofs was:
* To remind ourselves that there is more work to be done here - ideally this would be closed with `simp` or perhaps `simp with <some simp set>`
* To ensure that tother PRs don't undo the progress made here

If the `matrix.det ![![a, b], ![c, d]] = a * d - b * c` lemma was directly in mathlib, it would probably need to have the form
`matrix.det A = A 0 0 * A 1 1 - A 0 1 * A 1 0`.

But I think the idea was to leave this type of thing to a follow-up PR.

Reply:
A simp set can be constructed that is meant to unwrap sums as deep as possible. A lot of lemma would need to be tagged though for that simp set.